hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/server/SCMClientProtocolServer.java

```diff
@@ -261,6 +261,41 @@ public ContainerWithPipeline allocateContainer(HddsProtos.ReplicationType
     }
   }
 
+  @Override
+  public ContainerWithPipeline allocateContainer(ReplicationConfig replicationConfig, String owner) throws IOException {
+    Map<String, String> auditMap = Maps.newHashMap();
+    auditMap.put("replicationType", String.valueOf(replicationConfig.getReplicationType()));
+    if (replicationConfig.getReplicationType() == HddsProtos.ReplicationType.RATIS ||
+        replicationConfig.getReplicationType() == HddsProtos.ReplicationType.STAND_ALONE) {
+      auditMap.put("factor", String.valueOf(replicationConfig.getReplication()));
+    }
+    auditMap.put("owner", String.valueOf(owner));
+
+    try {
+      if (scm.getScmContext().isInSafeMode()) {
+        throw new SCMException("SafeModePrecheck failed for allocateContainer",
+            ResultCodes.SAFE_MODE_EXCEPTION);
+      }
+      getScm().checkAdminAccess(getRemoteUser(), false);
+      final ContainerInfo container = scm.getContainerManager()
+          .allocateContainer(
+              replicationConfig,
+              owner);
+      final Pipeline pipeline = scm.getPipelineManager()
+          .getPipeline(container.getPipelineID());
+      ContainerWithPipeline cp = new ContainerWithPipeline(container, pipeline);
+      AUDIT.logWriteSuccess(buildAuditMessageForSuccess(
+          SCMAction.ALLOCATE_CONTAINER, auditMap)
+      );
+      return cp;
+    } catch (Exception ex) {
+      AUDIT.logWriteFailure(buildAuditMessageForFailure(
+          SCMAction.ALLOCATE_CONTAINER, auditMap, ex)
+      );
+      throw ex;
+    }
```

Comment:
Please avoid duplication.  Extract a method for this, and let existing `allocateContainer` call it with `ReplicationConfig.fromProtoTypeAndFactor(replicationType, factor)`.